### PR TITLE
feat: rewrite bash installer script

### PR DESCRIPTION
**Pull request by jooola**
- rewrite the bash installer script
- remove old initd, sysv, apache<2, airtime migration steps
- use environment variables for advanced configuration alongside options flags
- new banner
- drop complex verbosity management (redirect stdout to /dev/null if that bother you)
- don't stop existing services during an upgrade procedure

### CODE REVIEW

**Comment by jooola** [comment]
This still needs some tweaks in the web installer to catch the details from the install script and remove some bad logic from the web installer.

But in the meantime I think this can already be reviewed.

**Comment by paddatrapper** [comment]
This needs an update to the install doc https://libretime.org/install changing the install command and explaining what environment variables are available.

**Comment by paddatrapper** [comment]
Getting the following errors trying to run this:

```
    ubuntu-focal:   ERROR: Command errored out with exit status 1:
    ubuntu-focal:    command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-w57soxp8/psycopg2/setup.py'"'"'; __file__='"'"'/tmp/pip-install-w57soxp8/psycopg2/setu
p.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-h8eul69a 
    ubuntu-focal:        cwd: /tmp/pip-install-w57soxp8/psycopg2/                                                                                                                                                     ubuntu-focal:   Complete output (38 lines):                                                                                                                                                                       ubuntu-focal:   running bdist_wheel                                                                                                                                                                           
    ubuntu-focal:   running build                                                                        
    ubuntu-focal:   running build_py                                                                                                                                                                              
    ubuntu-focal:   creating build                                                                                                                                                                                
    ubuntu-focal:   creating build/lib.linux-x86_64-3.8                                                                                                                                                           
    ubuntu-focal:   creating build/lib.linux-x86_64-3.8/psycopg2       
    ubuntu-focal:   copying lib/__init__.py -> build/lib.linux-x86_64-3.8/psycopg2                       
    ubuntu-focal:   copying lib/errorcodes.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/tz.py -> build/lib.linux-x86_64-3.8/psycopg2                                                                                                                                      
    ubuntu-focal:   copying lib/extensions.py -> build/lib.linux-x86_64-3.8/psycopg2                                                                                                                              
    ubuntu-focal:   copying lib/_range.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/errors.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/_ipaddress.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/pool.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/extras.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/_json.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   copying lib/sql.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:   running build_ext
    ubuntu-focal:   building 'psycopg2._psycopg' extension
    ubuntu-focal:   creating build/temp.linux-x86_64-3.8
    ubuntu-focal:   creating build/temp.linux-x86_64-3.8/psycopg
    ubuntu-focal:   x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector
-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DPSYCOPG_VERSION=2.9.3 (dt dec pq3 ext lo64) -DPSYCOPG_DEBUG=1 -DPG_VERSION_NUM=120009 -DHAVE_LO64=1 -DPSYCOPG_DEBUG=1 -I/usr/incl
ude/python3.8 -I. -I/usr/include/postgresql -I/usr/include/postgresql/12/server -I/usr/include/libxml2 -I/usr/include/mit-krb5 -c psycopg/psycopgmodule.c -o build/temp.linux-x86_64-3.8/psycopg/psycopgmodule.o -
Wdeclaration-after-statement
    ubuntu-focal:   In file included from psycopg/psycopgmodule.c:28:
    ubuntu-focal:   ./psycopg/psycopg.h:36:10: fatal error: libpq-fe.h: No such file or directory
    ubuntu-focal:      36 | #include <libpq-fe.h>
    ubuntu-focal:         |          ^~~~~~~~~~~~
    ubuntu-focal:   compilation terminated.
    ubuntu-focal: 
    ubuntu-focal:   It appears you are missing some prerequisite to build the package from source.
    ubuntu-focal: 
    ubuntu-focal:   You may install a binary package by installing 'psycopg2-binary' from PyPI.
    ubuntu-focal:   If you want to install psycopg2 from source, please install the packages
    ubuntu-focal:   required for the build and try again.
    ubuntu-focal: 
    ubuntu-focal:   For further information please check the 'doc/src/install.rst' file (also at
    ubuntu-focal:   <https://www.psycopg.org/docs/install.html>).
    ubuntu-focal: 
    ubuntu-focal:   error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ubuntu-focal:   ----------------------------------------
    ubuntu-focal:   Running setup.py clean for psycopg2
    ubuntu-focal:   ERROR: Failed building wheel for psycopg2

[...]

    ubuntu-focal:     ERROR: Command errored out with exit status 1:                                                                                                                                       [7/841]
    ubuntu-focal:      command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-w57soxp8/psycopg2/setup.py'"'"'; __file__='"'"'/tmp/pip-install-w57soxp8/psycopg2/se
tup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-veqv
dxda/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/include/python3.8/psycopg2
    ubuntu-focal:          cwd: /tmp/pip-install-w57soxp8/psycopg2/
    ubuntu-focal:     Complete output (40 lines):
    ubuntu-focal:     running install
    ubuntu-focal:     /usr/local/lib/python3.8/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
    ubuntu-focal:       warnings.warn(
    ubuntu-focal:     running build
    ubuntu-focal:     running build_py
    ubuntu-focal:     creating build
    ubuntu-focal:     creating build/lib.linux-x86_64-3.8
    ubuntu-focal:     creating build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/__init__.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/errorcodes.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/tz.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/extensions.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/_range.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/errors.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/_ipaddress.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/pool.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/extras.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/_json.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     copying lib/sql.py -> build/lib.linux-x86_64-3.8/psycopg2
    ubuntu-focal:     running build_ext
    ubuntu-focal:     building 'psycopg2._psycopg' extension
    ubuntu-focal:     creating build/temp.linux-x86_64-3.8
    ubuntu-focal:     creating build/temp.linux-x86_64-3.8/psycopg
    ubuntu-focal:     x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protect
or-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DPSYCOPG_VERSION=2.9.3 (dt dec pq3 ext lo64) -DPSYCOPG_DEBUG=1 -DPG_VERSION_NUM=120009 -DHAVE_LO64=1 -DPSYCOPG_DEBUG=1 -I/usr/in
clude/python3.8 -I. -I/usr/include/postgresql -I/usr/include/postgresql/12/server -I/usr/include/libxml2 -I/usr/include/mit-krb5 -c psycopg/psycopgmodule.c -o build/temp.linux-x86_64-3.8/psycopg/psycopgmodule.o
 -Wdeclaration-after-statement
    ubuntu-focal:     In file included from psycopg/psycopgmodule.c:28:
    ubuntu-focal:     ./psycopg/psycopg.h:36:10: fatal error: libpq-fe.h: No such file or directory
    ubuntu-focal:        36 | #include <libpq-fe.h>
    ubuntu-focal:           |          ^~~~~~~~~~~~
    ubuntu-focal:     compilation terminated.
    ubuntu-focal: 
    ubuntu-focal:     It appears you are missing some prerequisite to build the package from source.
    ubuntu-focal: 
    ubuntu-focal:     You may install a binary package by installing 'psycopg2-binary' from PyPI.
    ubuntu-focal:     If you want to install psycopg2 from source, please install the packages
    ubuntu-focal:     required for the build and try again.
    ubuntu-focal: 
    ubuntu-focal:     For further information please check the 'doc/src/install.rst' file (also at
    ubuntu-focal:     <https://www.psycopg.org/docs/install.html>).
    ubuntu-focal: 
    ubuntu-focal:     error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ubuntu-focal:     ----------------------------------------
    ubuntu-focal: ERROR: Command errored out with exit status 1: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-w57soxp8/psycopg2/setup.py'"'"'; __file__='"'"'/tmp
/pip-install-w57soxp8/psycopg2/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' inst
all --record /tmp/pip-record-veqvdxda/install-record.txt --single-version-externally-managed --compile --install-headers /usr/local/include/python3.8/psycopg2 Check the logs for full command output.
```

**Comment by jooola** [comment]
This isn't the first time I see this error and I am wonder if it does any harm:
```
    ubuntu-focal: info: upgrading python3 tools
    ubuntu-focal: Collecting setuptools~=58.0
    ubuntu-focal:   Downloading setuptools-58.5.3-py3-none-any.whl (946 kB)
    ubuntu-focal: ERROR: launchpadlib 1.10.13 requires testresources, which is not installed.
    ubuntu-focal: Installing collected packages: setuptools
    ubuntu-focal:   Attempting uninstall: setuptools
    ubuntu-focal:     Found existing installation: setuptools 45.2.0
    ubuntu-focal:     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    ubuntu-focal:     Can't uninstall 'setuptools'. No files were found to uninstall.
    ubuntu-focal: Successfully installed setuptools-58.5.3

```

**Comment by jooola** [comment]
@paddatrapper your error occur because the api packages aren't install because of some error with the tools/package.py tool.

I'll investigate
```
    ubuntu-focal: 
    ubuntu-focal: Traceback (most recent call last):
    ubuntu-focal:   File "/vagrant/tools/packages.py", line 113, in <module>
    ubuntu-focal:     run()
    ubuntu-focal:   File "/vagrant/tools/packages.py", line 104, in run
    ubuntu-focal:     packages = list_packages(args.path, args.distribution, args.dev, args.exclude)
    ubuntu-focal:   File "/vagrant/tools/packages.py", line 64, in list_packages
    ubuntu-focal:     for package_file in list_packages_files(paths):
    ubuntu-focal:   File "/vagrant/tools/packages.py", line 52, in list_packages_files
    ubuntu-focal:     raise Exception(f"{path} is not a file!")
    ubuntu-focal: Exception: packages.ini is not a file!

```

**Comment by paddatrapper** [comment]
> This isn't the first time I see this error and I am wonder if it does any harm:

Hmm... It doesn't seem like it is causing any issues and setuptools is upgraded... Maybe we need to look into the use of upgrading setuptools and pip to the latest version? The distro version will be supported by the distro and unless we are using some newer features of pip, maybe we can just use the default.

**Comment by paddatrapper** [comment]
> This isn't the first time I see this error and I am wonder if it does any harm:

Hmm... It doesn't seem like it is causing any issues and setuptools is upgraded... Maybe we need to look into the use of upgrading setuptools and pip to the latest version? The distro version will be supported by the distro and unless we are using some newer features of pip, maybe we can just use the default.

**Comment by jooola** [comment]
This PR will need #1559 and #1558

**Comment by jooola** [comment]
I think we can remove the `console` tool and use django custom commands to run some required installation tasks, such as creating the initial schema or storing CORS urls and such.

This prevent us from introducing an extra tool to the stack and those command might become handy with time.

The install process will then be:
[](https://mermaid.live/edit/#pako:eNqFU01rwzAM_SvCx9GssGMo3aU7DMYO-ziMZQQ3VlpBbAd_lI3S_z47SYNTmi2XGL2np2fJOrJKC2Q5s4473BDfGS6zw12hIHxdEMiWvg1xgbBaVXtNFa7XPeHz5guybJ1QcthotOD2CJVWNe2gpgYBv8k6C_eFSpUL9qjCqWngibYG30giWDSHUMEWDLgF6vFp0sbLFniqf4WcuO4c9lgOz3oq9iDIXdrt5TAgZR-dphxIwhJdteRkXPB8_t9G8kzuUL_zMgGn0i9edWbEFmy1RxluaZA70qrXNV6VPVCOwESgiY2MZjLeUqhKLhPbf3ITP52_WeIVYEyQFJodI_byTu_DGOZmPIzp6tjGV_WB9o9eJcXHq6aG5jvUs3A-LXVzDZ_GOlrYirgGCqPjzjVbMIlGchJh1Y4xsWDBugx183AUWHPfuIIV6hSovhXBUnyY2rC85o3FBePe6dcfVbHcGY9n0rCxA-v0C2OYUD8)


The config file check is made inside the installer.

**Comment by paddatrapper** [comment]
I like that method. Only comment is that the db-init step, as the Django migrations will do that given a DB and sufficiently privileged user

**Comment by jooola** [comment]
> I like that method. Only comment is that the db-init step, as the Django migrations will do that given a DB and sufficiently privileged user

Sorry, I don't think I understand.

**Comment by paddatrapper** [comment]
The `Run the DB schema creation` step is not required. Given a DB and user, Django will create and configure all the tables required when you run `migrate`

**Comment by jooola** [comment]
> The Run the DB schema creation step is not required. Given a DB and user, Django will create and configure all the tables required when you run migrate

This will only be true once the php migration are moved to django. Which is not in the scope of this PR I think. The DB schema creation is a Django command but runs code from legacy that is was initially run by the web installer. Or am I missing something ?

**Comment by paddatrapper** [comment]
I think we need to test - the Django migrations are run from the web installer because it needs the database connection details to be written into the configuration file so that it can read them from there. The only thing I'm unsure about is if Django will create tables in a database if they don't exist when `managed == False`. If it does not, then we'll need a init-schema step that will call the PHP required to configure those tables. If it will, then we just need to give it access to a DB and it will create the tables without a problem and no init-schema step is required.

**Comment by paddatrapper** [comment]
Ok, had a look at the docs - for unmanaged, it will not attempt to create them if they don't exist. So as long as we use Propel as the source of truth, we will need a init-schema step

File: console/libretime_console/setup.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                """
+                UPDATE cc_stream_setting
+                SET value = %(value)s
+                WHERE keyname = %(key)s
+                """,
+                {
+                    "key": "_".join(keys),
+                    "value": icecast_admin_password,
+                },
+            )
```
Comment (jooola) [inline console/libretime_console/setup.py:200]: This is the equivalent code the is run the web installer. I am currently working on moving this code to a django command

**Comment by jooola** [comment]
Closing in favor of #1821

**Pull request closed without merging**